### PR TITLE
Disallow `ref` to `in`/`ref readonly` method parameter conversion

### PR DESCRIPTION
**Pull request by jjonescz**
Fixes https://github.com/dotnet/roslyn/issues/71383.
Corresponding speclet update: https://github.com/dotnet/csharplang/pull/7820
Test plan: https://github.com/dotnet/roslyn/issues/68056

### CODE REVIEW

**Comment by jjonescz** [comment]
@microsoft-github-policy-service rerun

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs [hunk 1]
```diff
@@ -7180,6 +7180,19 @@ class C
         }
         else
         {
+            if (x == "ref" && y is "in" or "ref readonly")
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:7183]:
```suggestion
            if (x is "ref" && y is "in" or "ref readonly")
```

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs [hunk 2]
```diff
@@ -7240,19 +7233,57 @@ static class E
...
+        else
+        {
+            CompileAndVerify(new[] { source1, source2 }, expectedOutput: "X1 Y1 Y2", verify: Verification.FailsILVerify).VerifyDiagnostics(
+                // 1.cs(5,5): warning CS9198: Reference kind modifier of parameter 'in int p' doesn't match the corresponding parameter 'ref readonly int p' in target.
+                // x = c.Y1;
+                Diagnostic(ErrorCode.WRN_TargetDifferentRefness, "c.Y1").WithArguments("in int p", "ref readonly int p").WithLocation(5, 5),
+                // 1.cs(7,5): warning CS9198: Reference kind modifier of parameter 'in int p' doesn't match the corresponding parameter 'ref readonly int p' in target.
+                // x = c.Y2;
+                Diagnostic(ErrorCode.WRN_TargetDifferentRefness, "c.Y2").WithArguments("in int p", "ref readonly int p").WithLocation(7, 5));
+        }
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:7286]: Could we move this section directly after the declaration of `var source2` and before `var source3`? That way the results are closer to the source. Makes the test a bit easier to read.

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs [hunk 3]
```diff
@@ -7314,32 +7343,48 @@ class C
...
+                // x = (X)C.Y;
+                Diagnostic(ErrorCode.ERR_MethDelegateMismatch, "(X)C.Y").WithArguments("Y", "X").WithLocation(4, 5));
+        }
+        else
+        {
+            CompileAndVerify(new[] { source1, source2 }, expectedOutput: "XY").VerifyDiagnostics(
+                // 1.cs(4,5): warning CS9198: Reference kind modifier of parameter 'in int p' doesn't match the corresponding parameter 'ref readonly int p' in target.
+                // x = (X)C.Y;
+                Diagnostic(ErrorCode.WRN_TargetDifferentRefness, "(X)C.Y").WithArguments("in int p", "ref readonly int p").WithLocation(4, 5));
+        }
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:7381]: Can we move this section between the definition of `source2` and `source3`? That puts the code closer to the diagnostics and makes it a bit easier to read. 

Similar feedback for later tests.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs [hunk 4]
```diff
@@ -3322,28 +3322,34 @@ internal EffectiveParameters(ImmutableArray<TypeWithAnnotations> types, Immutabl
             return paramRefKind;
         }
 
-        // In method group conversions, 'in' is allowed to match 'ref' and 'ref readonly' is allowed to match 'ref' or 'in'.
-        internal static bool AreRefsCompatibleForMethodConversion(RefKind x, RefKind y, CSharpCompilation compilation)
+        // In method group conversions,
+        // - 'in' is allowed to be converted to 'ref',
+        // - 'ref readonly' is allowed to be converted to 'ref' or 'in',
+        // - 'in' is allowed to be converted to 'ref readonly'.
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:3328]: Reading through this chart, occurs to me that `out` should be allowed to convert to `ref`. I do **not** want to change the scope of the PR to include this. Just a thought that occured.

**Review by jaredpar — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs (hunk 4 above)
Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:3328]: Seems safe, but that probably needs a separate proposal, right?

Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:3328]: Agree. Not sure it's worth it, just occurred to me when reading this particular part of the change.

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 5]
```diff
@@ -1937,7 +1937,7 @@ internal bool MethodIsCompatibleWithDelegateOrFunctionPointer(BoundExpression? r
                 //  * Every ref or similar parameter has an identity conversion to the corresponding target parameter
                 // Note the addition of the reference requirement: it means that for delegate type void D(int i), void M(long l) is
                 // _applicable_, but not _compatible_.
-                if (!hasConversion(this, delegateType.TypeKind, Conversions, delegateParameter.Type, methodParameter.Type, delegateParameter.RefKind, methodParameter.RefKind, ref useSiteInfo))
+                if (!hasConversion(delegateType.TypeKind, Conversions, source: delegateParameter.Type, destination: methodParameter.Type, sourceRefKind: delegateParameter.RefKind, destinationRefKind: methodParameter.RefKind, checkingReturns: false, ref useSiteInfo))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:1940]: >if (!hasConversion(delegateType.TypeKind, Conversions, source: delegateParameter.Type, destination: methodParameter.Type, sourceRefKind: delegateParameter.RefKind, destinationRefKind: methodParameter.RefKind, checkingReturns: false, ref useSiteInfo)) [](http://example.com/codeflow?start=16&length=250)

It looks like this line became too long to fit an average screen. Consider splitting it

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 6]
```diff
@@ -1958,7 +1958,7 @@ internal bool MethodIsCompatibleWithDelegateOrFunctionPointer(BoundExpression? r
             bool returnsMatch = delegateOrFuncPtrMethod switch
             {
                 { RefKind: RefKind.None, ReturnsVoid: true } => method.ReturnsVoid,
-                { RefKind: var destinationRefKind } => hasConversion(this, delegateType.TypeKind, Conversions, methodReturnType, delegateReturnType, method.RefKind, destinationRefKind, ref useSiteInfo),
+                { RefKind: var destinationRefKind } => hasConversion(delegateType.TypeKind, Conversions, source: methodReturnType, destination: delegateReturnType, sourceRefKind: method.RefKind, destinationRefKind: destinationRefKind, checkingReturns: true, ref useSiteInfo),
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:1961]: >{ RefKind: var destinationRefKind } => hasConversion(delegateType.TypeKind, Conversions, source: methodReturnType, destination: delegateReturnType, sourceRefKind: method.RefKind, destinationRefKind: destinationRefKind, checkingReturns: true, ref useSiteInfo), [](http://example.com/codeflow?start=16&length=259)

Consider splitting this line too

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 7]
```diff
@@ -1992,10 +1992,13 @@ internal bool MethodIsCompatibleWithDelegateOrFunctionPointer(BoundExpression? r
...
-            static bool hasConversion(Binder binder, TypeKind targetKind, Conversions conversions, TypeSymbol source, TypeSymbol destination,
-                RefKind sourceRefKind, RefKind destinationRefKind, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+            static bool hasConversion(TypeKind targetKind, Conversions conversions, TypeSymbol source, TypeSymbol destination,
+                RefKind sourceRefKind, RefKind destinationRefKind, bool checkingReturns, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
             {
-                if (!OverloadResolution.AreRefsCompatibleForMethodConversion(sourceRefKind, destinationRefKind, binder.Compilation))
+                // Allowed ref kind mismatches between parameters have already been checked by overload resolution.
+                if (checkingReturns
+                    ? sourceRefKind != destinationRefKind
+                    : (sourceRefKind != RefKind.None) != (destinationRefKind != RefKind.None))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:2001]: >(sourceRefKind != RefKind.None) != (destinationRefKind != RefKind.None) [](http://example.com/codeflow?start=22&length=71)

```(sourceRefKind == RefKind.None) != (destinationRefKind == RefKind.None)```? It feels like there are too many ```!=``` in the code

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs [hunk 8]
```diff
@@ -1488,7 +1488,7 @@ private static LambdaConversionResult IsAnonymousFunctionCompatibleWithDelegate(
                 {
                     for (int p = 0; p < delegateParameters.Length; ++p)
                     {
-                        if (!OverloadResolution.AreRefsCompatibleForMethodConversion(delegateParameters[p].RefKind, anonymousFunction.RefKind(p), compilation) ||
+                        if (!OverloadResolution.AreRefsCompatibleForMethodConversion(source: anonymousFunction.RefKind(p), destination: delegateParameters[p].RefKind, compilation) ||
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs:1491]: >source: [](http://example.com/codeflow?start=85&length=7)

Naming of parameters feels confusing to me. Data flow from delegate parameter to the lambda parameter. So, I would expect the delegate parameter to be the "source".

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs [hunk 9]
```diff
@@ -3305,7 +3305,7 @@ internal EffectiveParameters(ImmutableArray<TypeWithAnnotations> types, Immutabl
                     return argRefKind;
                 }
             }
-            else if (AreRefsCompatibleForMethodConversion(paramRefKind, argRefKind, binder.Compilation))
+            else if (AreRefsCompatibleForMethodConversion(source: paramRefKind, destination: argRefKind, binder.Compilation))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:3308]: >source [](http://example.com/codeflow?start=58&length=6)

Similar comment here about parameter naming. Argument is the "source", its value flows to the parameter

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs [hunk 10]
```diff
@@ -3322,28 +3322,34 @@ internal EffectiveParameters(ImmutableArray<TypeWithAnnotations> types, Immutabl
             return paramRefKind;
         }
 
-        // In method group conversions, 'in' is allowed to match 'ref' and 'ref readonly' is allowed to match 'ref' or 'in'.
-        internal static bool AreRefsCompatibleForMethodConversion(RefKind x, RefKind y, CSharpCompilation compilation)
+        // In method group conversions,
+        // - 'in' is allowed to be converted to 'ref',
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:3326]: >'in' is allowed to be converted to 'ref', [](http://example.com/codeflow?start=13&length=41)

Same comment as for the speclet update. I don't think the language has a concept conversion for modifiers.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs [hunk 11]
```diff
@@ -3322,28 +3322,34 @@ internal EffectiveParameters(ImmutableArray<TypeWithAnnotations> types, Immutabl
...
+        // - 'ref readonly' is allowed to be converted to 'ref' or 'in',
+        // - 'in' is allowed to be converted to 'ref readonly'.
+        internal static bool AreRefsCompatibleForMethodConversion(RefKind source, RefKind destination, CSharpCompilation compilation)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:3329]: >source [](http://example.com/codeflow?start=74&length=6)

As mentioned earlier, I find the naming of parameters confusing. Would it make sense to switch to terms like "delegateParameterRefKind" and "candidateMethodParameterRefKind"?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 4), tests are not looked at

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 5). I didn't do thorough review of test, rely on @jaredpar sign-off for that.

**Comment by jaredpar** [comment]
> I didn't do thorough review of test, relied on @jaredpar sign-off for that.

Primarily reviewed those, made sure they lined up with spec and eexpectations.

**Pull request merged**